src/payment/strategies/cardinal/cardinal-three-d-secure-flow-v2.ts

```diff
@@ -0,0 +1,111 @@
+import { merge, some } from 'lodash';
+
+import { CheckoutStore, InternalCheckoutSelectors } from '../../../checkout';
+import { RequestError } from '../../../common/error/errors';
+import { HostedForm } from '../../../hosted-form';
+import { OrderPaymentRequestBody, OrderRequestBody } from '../../../order';
+import { InstrumentSelector } from '../../instrument';
+import isCreditCardLike from '../../is-credit-card-like';
+import isVaultedInstrument from '../../is-vaulted-instrument';
+import PaymentActionCreator from '../../payment-action-creator';
+import PaymentMethod from '../../payment-method';
+import { PaymentRequestOptions } from '../../payment-request-options';
+import PaymentStrategy from '../payment-strategy';
+
+import { CardinalThreeDSecureToken } from './cardinal';
+import CardinalClient, { CardinalOrderData } from './cardinal-client';
+
+export default class CardinalThreeDSecureFlowV2 {
+    constructor(
+        private _store: CheckoutStore,
+        private _paymentActionCreator: PaymentActionCreator,
+        private _cardinalClient: CardinalClient
+    ) {}
+
+    async prepare(method: PaymentMethod): Promise<void> {
+        await this._cardinalClient.load(method.id, method.config.testMode);
+    }
+
+    async start(
+        execute: PaymentStrategy['execute'],
+        payload: OrderRequestBody,
+        options?: PaymentRequestOptions,
+        hostedForm?: HostedForm
+    ): Promise<InternalCheckoutSelectors> {
+        const { instruments: { getCardInstrument } } = this._store.getState();
+        const { payment = { methodId: '' } } = payload;
+        const { paymentData = {} } = payment;
+
+        try {
+            return await execute(payload, options);
+        } catch (error) {
+            if (error instanceof RequestError && error.body.status === 'additional_action_required') {
+                const token = error.body.additional_action_required?.data?.token;
+                const xid = error.body.three_ds_result?.payer_auth_request;
+
+                await this._cardinalClient.configure(token);
+
+                const bin = this._getBin(paymentData, getCardInstrument, hostedForm);
+
+                if (bin) {
```

Comment:
what happens if there's no bin? should we at least throw an exception?

Reply:
As per documentation: _"Failure to run BIN Detection may result in higher step up rates and/or downgrades to 3DS 1.0.2"_ So, I think that's not enough to interrupt the flow.

Reply:
Also, I think it would be extremely rare not to have a bin. 🤔